internal/blocksync/reactor.go

```diff
@@ -185,31 +185,34 @@ func (r *Reactor) OnStop() {
 // Otherwise, we'll respond saying we do not have it.
 func (r *Reactor) respondToPeer(ctx context.Context, msg *bcproto.BlockRequest, peerID types.NodeID, blockSyncCh *p2p.Channel) error {
 	block := r.store.LoadBlock(msg.Height)
-	if block != nil {
-		extCommit := r.store.LoadBlockExtendedCommit(msg.Height)
-		if extCommit == nil {
-			return fmt.Errorf("found block in store without extended commit: %v", block)
-		}
-		blockProto, err := block.ToProto()
-		if err != nil {
-			return fmt.Errorf("failed to convert block to protobuf: %w", err)
-		}
-
+	if block == nil {
+		r.logger.Info("peer requesting a block we do not have", "peer", peerID, "height", msg.Height)
 		return blockSyncCh.Send(ctx, p2p.Envelope{
-			To: peerID,
-			Message: &bcproto.BlockResponse{
-				Block:     blockProto,
-				ExtCommit: extCommit.ToProto(),
-			},
+			To:      peerID,
+			Message: &bcproto.NoBlockResponse{Height: msg.Height},
 		})
 	}
-
-	r.logger.Info("peer requesting a block we do not have", "peer", peerID, "height", msg.Height)
+	extCommit := r.store.LoadBlockExtendedCommit(msg.Height)
+	if extCommit == nil {
+		c := r.store.LoadBlockCommit(msg.Height)
```

Comment:
I think we can simplify this part.
If the height requires extensions, we check the store for the extended commit. If it's not there, we don't send the message (just like in #8433).
If the height don't not require extensions, then we just ship `nil` as extension, as the receiver doesn't need them anyway.
The logic upon reception, can be simplified in the same way.
This way we avoid sending commits around in all heights before the "required height".

Reply:
To adapt the receiver logic, you just need to add a check for `VoteExtensionsEnabled(height)` to the existing one in line 249 below.

Reply:
Done on the `respondToPeer` side.

When receiving from the peer, I think it is slightly wrong to do this check. We may receive blocks way ahead of our current state. The state may have changed between the height we know about and the height we just received. If vote extensions were enabled in that time, we'll erroneously reject them (`ABCI.VoteExtensionsEnableHeight == 0` means reject). We need to validate the block in the presence of the `state` object that corresponds to their height.

Reply:
> Also, after thinking about it a bit, I'm really glad we decided to not allow extensions to be deactivated after activating them. I realize supporting that would have been a nightmare in this logic here.

Yeah, I definitely agree. It could certainly have been done, but it would have been much messier here.

> One last thought, though. I believe that, for the case of lines 203-206, we can simply ship nil as extension, rather than the stripped commit. Here's the reasoning: if I am being requested a block in my past, and I don't have the corresponding extended commit in my store, I am sure that the consensus param was not activated for that height. Also, sending a commit as extended commit in that case doesn't contain any useful info for the receiver.

I believe that what you're describing is what we do here. On line 198, we wrap the commit with missing extensions in an extended commit where the extension fields are nil.

Reply:
What mean is that. If you query the block store and don't have an extended commit, you _know_ that the peer asking for that height won't need that either (it'll have it through other means). So, here, you don't need to include _any_ commit in the message

Reply:
> What mean is that. If you query the block store and don't have an extended commit, you know that the peer asking for that height won't need that either (it'll have it through other means). So, here, you don't need to include any commit in the message

I'm not sure I understand, we are sending it height `H`, won't it need the commit for `H` as well? The block includes the commit for `H-1`, but not the commit for `H`. To begin `H+1`, i.e. to switch to consensus, we need to have the commit for height `H` both to believe it was actually committed by the network _and_ in case we propose `H+1`.

Reply:
What you said is correct. However, since we always process the blocks two by two, we'll always have the needed commit from `second`, so no need for the extended commit.
It is actually how it was working originally, before #8433; after "peeking two blocks" we were feeding the block store with `second.LastCommit`, pretending it was a `seenCommit`

Reply:
Got it, I just gave this a try. This requires adding conditional checks in many places throughout the blocksync code and storing / checking nils throughout the code. It ultimately saves a few bytes of data on the wire and in the store. I think this may be worth doing as an optimization, but I'd prefer to come back and clean it up. Once we know everything is working

Reply:
Yeah, that seems like a good plan to me